### ISSUE TITLE
Assembly.GetType doesn't support AssemblyQualified names.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/runtime#56776]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/AssemblyGetType-doesnt-support-Assembl/1492976)._

---
[severity:It's more difficult to complete my work]
Assembly.GetType doesn't support AssemblyQualified names. Here's an example:

var aqn = <an assembly qualified name of a type in myAssembly>; 
foreach (var assemblyType in myAssembly.GetTypes())
{
    if (assemblyType.AssemblyQualifiedName == aqn)
    {
        var test = myAssembly.GetType(assemblyType.AssemblyQualifiedName);
        **//test is null here! this is the issue. I'm directly passing the name from the assembly's type back into itself and it can't find the type.**

test = myAssembly.GetType(assemblyType.FullName);
        //test is no longer null. It can find the type with the FullName, but not the AssemblyQualifiedName.

return assemblyType;
    }
}

---
#### Feedback Bot on 8/2/2021, 10:52 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by GrabYourPitchforks** [comment]
I'm inclined not to change the behavior here. This method is clearly documented as "takes the FullName, not the AssemblyQualifiedName." Both the sample code and the remarks section of the docs call out this behavior.

**Comment by ryank425** [comment]
I think OP is confused with two different methods.

Assembly.GetType(string) expects you to have "The full name of the type." https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettype?view=net-5.0#System_Reflection_Assembly_GetType_System_String_

Type.GetType(string) expects you to have "The assembly-qualified name of the type to get." https://docs.microsoft.com/en-us/dotnet/api/system.type.gettype?view=net-5.0#System_Type_GetType_System_String_

OP is calling myAssembly.GetType(...) which I assume myAssembly to be some Assembly object, and fullName is the expected. Type.GetType(...) expects the assembly qualified name.

**Comment by bcreswell** [comment]
Yes Type.GetType() accepts an assembly qualified name and that's why I assumed Assembly.GetType() would as well. The trouble is if you get the FullName of a generic type, the generic argument will be represented as an assembly qualified name even though you asked for FullName. 

Example:
var fullName = (new List<string>()).GetType().FullName;
The fullName variable here is "System.Collections.Generic.List`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]"

That representation of the inner generic argument for "string" is an assembly qualified name. If i want to do the reverse operation for a type in a given assembly, currently i have to:
1) Parse the type name string and extract the inner assembly qualified name of the generic argument.
2) Iterate through all the types in the assembly looking for a type with a matching assembly qualified name.

This is a mountain of work just to do the inverse operation of GetType().FullName. And the parsing gets messy depending on how many generic arguments the generic type has. Right now i'm only letting my users use List<TheirType> to keep it simple.

For context: i have a plugin architecture where other groups in the company can publish dlls that are loaded at runtime by my client app.

**Comment by bcreswell** [comment]
That workaround isn't going to work i dont think. Consider this serialization use case:


1. Use AssemblyType.GetType().FullName to get the type's string
2. store that string to disk
3. later, try to recreate the type using the string. 

We wont have the type object available. Wont be able to call "IsGenericType" on it because all we have is the string. The whole point of Assembly.GetType() is to get the type based on a string when you dont have the type object. The end result is I still have to write a very complicated GetType() method of my own because the one in .net doesn't work. 

Please reconsider. 

We store the FullName of the type but the FullName of the type isn't enough for Assembly.GetType() to recreate it because the assembly qualified name is a component of the full name. 

1. Assembly.GetType() asks for the full name of the type
2. An assembly qualified name can be a component of the Full name
3. Assembly.GetType() doesn't accept that assembly qualified name component of the full name
4. therefore, Assembly.GetType() doesn't work correctly. **It doesn't accept all full names** because it wont accept the full name of a generic.

This is clearly a Bug with .net that should be addressed.

**Comment by GrabYourPitchforks** [comment]
The following sample app works and prints the correct results, demonstrating that `Assembly.GetType(string)` works with the output from `Type.FullName`, even when the type's full name includes the assembly-qualified name of generic parameters.

Am I misunderstanding something about this bug report?

```cs
using System;
using System.Collections.Generic;

namespace app
{
    class Program
    {
        static void Main(string[] args)
        {
            var asm = typeof(List<>).Assembly;
            Console.WriteLine(asm);
            var typeFullName = typeof(List<string>).FullName;
            Console.WriteLine(typeFullName);
            var type = asm.GetType(typeFullName);
            Console.WriteLine(type);
        }
    }
}
```

Output:

```txt
System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
System.Collections.Generic.List`1[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
System.Collections.Generic.List`1[System.String]
```

**Comment by buyaa-n** [comment]
> Am I misunderstanding something about this bug report?

You are right, i thought @bcreswell is saying that did not work, now I don't know what is the issue here.

**Comment by bcreswell** [comment]
i made a boiled down example. 1 class library with a single class in it:
```
public class UserType
{
    public string Blah;
}
```

and a second project that loads the first project's assembly at runtime using reflection. It creates a list of the UserType and attempts to serialize the type to a string using Type.FullName and then deserialize the type using Assembly.GetType()

```
string assemblyFilePath = @"..\..\..\UserTypeAssembly\bin\Debug\UserTypeAssembly.dll";

var userAssembly = Assembly.LoadFrom(assemblyFilePath);
Type theirType = userAssembly.GetTypes().FirstOrDefault();

Type genericListType = typeof(List<>);
Type fullListType = genericListType.MakeGenericType(theirType);
var listOfUserTypeName = fullListType.FullName;

Console.WriteLine(listOfUserTypeName);

var resultType = userAssembly.GetType(listOfUserTypeName);
//resultType is null here
```
```
System.Collections.Generic.List`1[[UserTypeAssembly.UserType, UserTypeAssembly, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]
```

Using the debugger, i can see the resultType variable at the end is null. This serialization use case is very common and important for Assembly.GetType(). Type.AssemblyQualifiedName wasn't used explicitly.

**Comment by GrabYourPitchforks** [comment]
@bcreswell Your call to `userAssembly.GetType` near the end of the sample is using the wrong assembly. You should be calling this API on the assembly that contains `List<T>`, not on the assembly that contains `UserType`. This is because the string you're passing in is `listOfUserTypeName = "System.Collections.Generic.List<...>"`.

**Comment by bcreswell** [comment]
I've modified the example below. Same result. It's still null.
```
string assemblyFilePath = @"..\..\..\UserTypeAssembly\bin\Debug\UserTypeAssembly.dll";

var userAssembly = Assembly.LoadFrom(assemblyFilePath);
Type theirType = userAssembly.GetTypes().FirstOrDefault();

Type genericListType = typeof(List<>);
Type fullListType = genericListType.MakeGenericType(theirType);

//System.Collections.Generic.List`1[[UserTypeAssembly.UserType, UserTypeAssembly, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]
Console.WriteLine(fullListType.FullName);

//mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
Console.WriteLine(fullListType.Assembly);

var resultType = fullListType.Assembly.GetType(fullListType.FullName);
//resultType variable is null
```
